src/NodeHandler.ts

```diff
@@ -0,0 +1,55 @@
+import { Evented } from '@dojo/core/Evented';
+import { VNodeProperties } from '@dojo/interfaces/vdom';
+import Map from '@dojo/shim/Map';
+import { NodeHandler as NodeHandlerInterface } from './interfaces';
+
+/**
+ * Enum to identify the type of event.
+ * Listening to 'Projector' will notify when projector is created or updated
+ * Listening to 'Widget' will notifiy when widget root is created or updated
+ */
+export enum Type {
+	Projector = 'Projector',
+	Widget = 'Widget'
+}
+
+export class NodeHandler extends Evented implements NodeHandlerInterface {
+
+	private _nodeMap = new Map<string, HTMLElement>();
+
+	public get(key: string): HTMLElement | undefined {
+		return this._nodeMap.get(key);
+	}
+
+	public has(key: string): boolean {
+		return this._nodeMap.has(key);
+	}
+
+	public add(element: HTMLElement, properties: VNodeProperties) {
+		const key = String(properties.key);
+		this._nodeMap.set(key, element);
+		this.emit({ type: key });
+	}
+
+	public addRoot(element: HTMLElement, properties: VNodeProperties) {
```

Comment:
I know this is a bit picky, but you could have a single function that does this, and three specific functions that pass their type. Like:

```ts
public add(element: HTMLElement, properties: VNodeProperties) {
    this._add(element, properties, Type.Widget);
}

public addRoot(element: HTMLElement, properties: VNodeProperties) {
    this._add(element, properties, Type.Widget);
}

public addProjector(element: HTMLElement, properties: VNodeProperties) {
    this._add(element, properties, Type.Projector);
}

private _add(element: HTMLElement, properties: VNodeProperties = {}, type: Type) {
    if (properties.key || Type.Widget) {
        this.add(element, properties);
    }
    this.emit({ type });
}
```

Reply:
fair

Reply:
Isn't that logically actually same?

Reply:
disclosure, I updated the code :)